File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
   
        visited = [[0 for _ in range(len(grid))] for _ in range(len(grid))]
        queue = deque()
        queue.append([0, 0, 1])
        
        while len(queue) > 0:
            x, y, currLength = queue.popleft()
            
            if x < 0 or x >= len(grid):
                continue
            if y < 0 or y >= len(grid):
                continue
            if visited[x][y] == 1:
                continue
            if grid[x][y] == 1:
                continue
            if x == len(grid) - 1 and y == len(grid) - 1:
                if grid[x][y] == 0:
                    return currLength
                continue
                
            visited[x][y] = 1
                
            for i in range(x-1, x+2):
                for j in range(y-1, y+2):
                    queue.append([i, j, currLength+1])
                    
        return -1
                

File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.java
class Solution {
    public int shortestPathBinaryMatrix(int[][] grid) {
        int n = grid.length;
        if(grid[0][0] == 1 || grid[n - 1][n - 1] == 1) return -1;
        int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0},{1,1},{-1,-1},{-1,1},{1,-1}};
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{0,0});
        int res = 0;
        while(!q.isEmpty()){
            int size = q.size();
            res++;
            while(size-- > 0){
                int[] cur = q.poll();
                int x = cur[0], y = cur[1];
                if(x == n - 1 && y == n - 1) return res;
                for(int[] dir : dirs){
                    int nx = x + dir[0], ny = y + dir[1];
                    if(nx < 0 || nx >= n || ny < 0 || ny >= n || grid[nx][ny] == 1) continue;
                    grid[nx][ny] = 1;
                    q.add(new int[]{nx,ny});
                }
            }
        }
        return -1;
    }
}